TextureManagerクラスのリファクタリングと修正

Moto0116 opened the pull request:
**修正内容**
- vector関係のkeyという名前の変数はindexに修正した
- index初期化用の定数を作成
- 読み込みに失敗したらNULLへのkeyを返すように修正
- ReleaseTextureではNULLチェックを行うようにした

issue #149

masami-sugao commented on src/KingdomCraft/KingdomCraft/Library/TextureManager/TextureManager.cpp:
```diff
@@ -6,19 +6,21 @@
 #include <d3dx11.h>
 
-
+const int TextureManager::m_DefaultIndex = 0;
```
名前をもう少しわかりやすくしたいですね。
「DefaultIndex」だと初期値という印象を受けますが、無効なインデックスを示すのに使うんですよね？
そうなると、Invalidとかかなー。

Moto0116 replied:
確かにその通りです。
`m_InvalidIndex`に修正しておきます。

Moto0116 replied:
変数名`m_DefaultIndex`を`m_InvalidIndex`に修正しました。

masami-sugao commented on src/KingdomCraft/KingdomCraft/Library/TextureManager/TextureManager.cpp:
```diff
@@ -6,19 +6,21 @@
+TextureManager::TextureManager(ID3D11Device* const _pDevice) :
 m_pDevice(_pDevice)
 {
+	m_pTextureResourceView.push_back(NULL);
```
パッと見ではどういう意図でNULLを入れているのかわかりにくいので、コメント書いてある方がいいですね。

Moto0116 replied:
確かにコードを読まないとわからない状態でした。
コメントを付けて修正します。

Moto0116 replied:
コンストラクタの処理に対してコメントを追加して修正しました。

masami-sugao commented on src/KingdomCraft/KingdomCraft/Library/TextureManager/TextureManager.h:
```diff
@@ -31,46 +30,48 @@ class TextureManager
 	 * @return テクスチャの読み込みに成功したらtrue
 	 */
-	bool LoadTexture(LPCTCH _filePath, int* _pkey);
+	bool LoadTexture(LPCTSTR const _pFileName, int* const _pIndex);
```
呼び出し側でこの関数の成否を`_pIndex`のみで判断できるのであれば戻り値はvoidでよい気がします。
ただ、「ライブラリなので戻り値でもout引数でもどちらでも判断可能なようにしてます」というスタンスも有りだと思うので、そういう思想であればこのままで問題ないです。

Moto0116 replied:
自分としては呼び側で`if(!LoadTexture())`って判断できたほうが見やすいと思ったので返り値付をつけています。

taishi-matsumura commented on src/KingdomCraft/KingdomCraft/Library/TextureManager/TextureManager.cpp:
```diff
@@ -39,18 +41,29 @@ bool TextureManager::LoadTexture(LPCTCH _filePath, int* _pkey)
 	return true;
 }
+
+void TextureManager::ReleaseTexture(const int _index)
```
`int _index`にconst付けるのって、今日学校で言ってたやつよね？
うーん、さすがにウザくない？？

Moto0116 replied:
すみません、`_index`に関してはむしろつけても良いと思ったんですけど何か問題あったりしますか？

Moto0116 replied:
確かにその通りだと思います。
でも、ライブラリ部分に関しては、￥そんなに機能がたくさん追加されたり、頻繁に変更がかかったりしないので問題ないと思います。

masami-sugao replied:
私の考えを書いておきますね。
Webで見ても色々意見があるようなので、あくまで1プログラマーの意見として見て下さい。

ソースコードは可能な限りシンプルであるべきと思っています。
そして、各プロジェクトに求められる処理効率、可読性、保守性、拡張性、開発効率等を損なわない範囲で最も無駄のないプログラムを書くのがプログラマーに求められることだと。

意味がないことが自明である値渡しされた引数の書き換えをわざわざ文法で縛っているとなると、きっと何か特別な意味があるのだろうと読み手は考えるんです。
命名規則で引数にアンダースコアを付けて、変数と勘違いして値を書き換えることを予防しているので余計にそう考えます。
（そもそも処理内でそれが引数だということがわからなくなるような長い処理は関数を分けるべきだと思いますが）
それを各所で考えさせられるという意味で「読み辛い」です。
見た目に煩いのもありますが。

…という感じかなぁ。
ごめんなさい、私も上手く表現できている自信は全くないですが、読んだときモヤモヤしました。

Moto0116 replied:
貴重な意見ありがとうございます。
参考にさせていただきます。

motolz committed:
[issue #149] TextureManagerクラスのリファクタリングと修正

- vector関係のkeyという名前の変数はindexに修正した
- index初期化用の定数を作成
- 読み込みに失敗したらNULLへのkeyを返すように修正
- ReleaseTextureではNULLチェックを行うようにした
- .hのけつに改行追加
- メンバの定義順が間違えていたので修正
- コンストラクタ内の処理にコメント追加
- 関数の引数に付けるconstを一部外した

Moto0116 commented:
コミットログをまとめました。
内容に問題なければ合流お願いします。
